### ISSUE TITLE
More readable party chat - hide skills log

### ISSUE THREAD

**Issue by rcichy** [HabitRPG/habitica#3553]
Party chat is often hard to read because of spell casting.
The idea is to hide unnecesary spells information in party chat.

My party handle this feature this way:
<code>$('li').has('code').hide()</code>

I think about it as a toggle button, which stays toggled even if you log from another browser on another day.

UI proposal:
![capture2](https://cloud.githubusercontent.com/assets/5044995/3144173/25142e5e-ea1a-11e3-85a2-da7c4fd00896.PNG)
##

**Comment by deilann** [comment]
This isn't just going to hide spells -- it's going to hide anything that has `code` in it. That includes boss damage and anything a user types with `code`. I think this is a poor implementation and unnecessary.

**Comment by zingbretsen** [comment]
Ideally, battle notifications would be separate from the chat. Maybe on different tabs within the party window, or in separate frames?

As it is, chatting pretty much stops during quests because all actual messages get displaced by waves of spells.

**Comment by rcichy** [comment]
@deilann I agree. I don't thought about using above jQuery code to do this.
I thought about adding a css class to chat talk/spell cast and then filtering by toggling button.

**Comment by deilann** [comment]
@zingbretsen Are y'all seriously doing more than 200 spells a day?

**Comment by zingbretsen** [comment]
@deilann ... We cast a lot of spells. Maybe not 200/day, but a lot. 

We're a large party, some high level mages, lots of MP to go around. 

Regardless, I think it makes sense to either physically separate status updates from actual messages or include some kind of toggle to do so. As it stands, the code tags were the only thing that really seemed to differentiate spells from messages, and I've never seen anyone include a code tag in our party chat. So the snippet from @yafinee works for me for now.

But if the status updates could be assigned a class, it would be pretty easy to add a toggle, no?

**Comment by deilann** [comment]
Mmm, parties are designed for smaller groups, not larger groups is probably
a big part of your issue.

This would definitely be low priority, and probably better served with a
user script than building it into the site, which would create legacy that
aids a very small number of users.

**Comment by LB7979** [comment]
Yes all the Boss and Spell messages obscure the real Party chat, I agree with yafinee and zingbretsen that we should be able to hide that.
I'm in a 17-member Party that is continuously questing and generous in using Spells, and 95% of what is seen in Party chat is Boss and Spell messages.
For reference: we can look back up to 4 days in Chat, so we generate about 50 messages a day on average. I don't think that's really that extreme much as deilann is saying. I was in a 40+ member Party before, and there you would miss what your Party members would say if you didn't read Party Chat at least once a day.
As to users using `code` in their chat: how much is that really used? I've seen it in the Tavern sporadically, but never in my Party Chat.
And if there would be a code-hiding feature in HRPG (**optional** of course!), Party members would _know_ messages in `code` would not show up, so they wouldn't use `code` and it wouldn't be a problem.

**Comment by Alys** [comment]
There's some good discussion about this issue's requirements in an incomplete pull request and it should be read, as well as the points above, by anyone who wants to work on this: https://github.com/HabitRPG/habitrpg/pull/5364 

Some points of my own:
- We should not implement this by hiding all instances of the `<code>` tag. Many players do use `<code>` in their messages and they should not have to avoid that.
- It should remain possible to view human chat messages and system chat messages together in chronological order to allow players to comment on them (_"Hey StormBringer, thanks so much for those Blessings you just cast!"_). I.e., human and system chat messages should not be split into two separate streams.
- Allowing players to hide skill messages is a good idea. I'm not sure that hiding quest progress messages is good - at the moment, I'm opposed to that. The primary function of parties and quests is to provide accountability for your tasks, and the quest progress messages are an indication of how much you've accomplished or missed; allowing them to be hidden reduces the accountability. Hiding them is also likely to increase player's confusion about why they've taken damage (not an uncommon point of confusion already). Also, the quest progress messages are likely to be less spammy than skill casts.
- At the moment, I'm in favour of a temporary toggle that you have to re-select every time you want to hide skill messages, so that you have at least some visibility of how much your party members are helping you with skills. I understand that there'll be other points of view about this though. :)
- Hiding only skills and not quest progress messages would be made easier by giving them each a different "user" name, or have some other attribute to make them different. We should bear in mind though that changes to the API and database structure are best avoided at the moment unless essential, while a new version of the API is being written by our senior devs, so this _might_ have an impact on how much we can do with this.

**Comment by hwoodward** [comment]
I think the ability to hide quest progress is important - in even a middling party you get a bunch of them in a wave, and they are often less useful than spell messages - its easier to see health of the boss and your party than what buffs have already been cast. I don't know about making it automatic or not. 

With regards to keeping everything in one chat - my party has non-quest messages in our chat that we really want to be able to see without the quest messages. Separating the two completely would be a great solution for our, and likely many other, groups.

**Comment by hwoodward** [comment]
Also, if a 'username' is used to differentiate the messages for hiding, it makes a lot of sense to put all the damage and quest progress messages under a user named for the quest, since there specifically related to the quest, while spells happen even when there isn't a quest.

**Comment by Alys** [comment]
@hwoodward Fair point about hiding quest progress messages - I can see they would be bulky in a large party, and you're right that it should be possible to view chat messages without seeing either the quest or skill messages. However for quest messages, I'm even more in favour of a toggle that's temporary so that you do always have some visibility of them (for the reasons of accountability and avoiding confusion about damage that I mentioned above).

If we do provide a way to hide both quest and skill messages, then I don't think there's any reason to separate them completely from the player chat messages, and doing so would disadvantage the parties that do have conversations about the quest/skill messages.

**Comment by lynxlynxlynx** [comment]
I would just use less obtrusive colours for these messages. The box has the same highlight as when you're directly mentioned and the red of the line and text stand out a lot. It's the reverse of the way it should be.

**Comment by hwoodward** [comment]
I don't know, they don't have a user attribution, so they don't really look like someone posted them, and the direct mention is correct, because they directly impact you due to either damage possibility or stat alteration.

**Comment by hwoodward** [comment]
@Alys 
I agree that if we allow them to be hidden keeping them in the timeline works - it allows for more options. I don't think it would be bad to have the toggle temporary.

**Comment by librarianmage** [comment]
Should this be put on hold?

**Comment by Alys** [comment]
@MathWhiz It's suggestion-discussion so it's not currently open for changes to be made, and we're still discussing it intermittently so I think leaving it as suggestion-discussion is probably best. We won't be marking it as help welcome until after the website improvement project is finished.

**Comment by rakheepatel** [comment]
Thanks @yafinee 
With your suggestion I went ahead and made this bookmarklet. It's not perfect, but it does the job. For anyone else who comes here looking for this solution, see image on how to make your own bookmarklet. Pop the following code in the *URL* field:
 `javascript:(function(){$('li').has('code').hide()})();`


![hrpg_hide_sys_msg](https://user-images.githubusercontent.com/5573174/30683354-7e66b2c2-9ef0-11e7-93c2-a3222433c8c0.png)

**Comment by librarianmage** [comment]
Has there been any internal decisions on this? Personally, I think it would be easier if system messages weren't in a message box, but were instead outside of a chat box with smaller text. Messaging apps that I use tend to do that with "administrative" messages (adding new people, changing the title, etc).

**Comment by Alys** [comment]
I'm still really keen to keep system messages in normal party chat, in chronological order with all the messages from players, and with the system messages never hidden by default, but hideable with a button you'd click each time you visit party chat. This has the best chance of avoiding confusion from players about why they've suffered loss of health or why their stats have improved.

I like MathWhiz's idea of system messages not being in a normal message box, especially now that messages take up so much space. They could stretch across the whole width of the chat area since there's no avatar for them. They would still need the datestamp and the "like" button and the count of likes. They would not need the report button and that's already being removed in a different issue.

I _think_ there is an intention in the future to allow a skill to be cast multiple times with one click (similar to how you can now buy multiple items with one purchase event), and when/if that happens, it will help with this problem - it will be easy to insert a system message saying "Bob casts Ethereal Surge 10 times".

**Comment by LunarGem** [comment]
I definitely think this is important now especially with the new UI with all of the scrolllllling. (Are we dealing with all of the scrolling? It probably would be even worse on mobile.) Maybe like in Google Hangouts where the text is tiny and grey?

**Comment by lemoness** [comment]
I can confirm that in the future there will be the option for batch skill
casting, but it isn't the highest item on our priority list right now.

This is a good topic for @Tressley to look over.

**Comment by melarish** [comment]
@rakheepatel Sadly that didn't work for me :/ And either way, it would require showing the Bookmarks Bar which would waste screen space.

I'm really keen to see this feature added! I've completely given up on party chat. Ours is a small party, not very active at chatting but super active questing and buffing. I check the chat when I remember, and every time it's a long scroll through system messages to find if a human has said anything :( Which is probably why our party chat is so sad and lacking. I really want to see Habit have a social feature where you really get to know a few people well (which is much harder in a 1000+ member guild or 1,000,000+ member Tavern).

I've seen this requested a lot on Trello over the years (see the 2nd most recent comment on https://trello.com/c/LlJccgmY/293-chat-tavern-talk-enhancements-party-talk-enhancements-friend-list). Was really hoping I could contribute to it!

Slightly related question: is there any ticket/planned feature of highlighting unread messages? (didn't find any searching for 'unread' in Github but it's mentioned on the Trello card)

**Comment by rakheepatel** [comment]
@melarish Yeah that only worked for the old layout. Since the site revamp, it'll no longer work.

**Comment by texastoland** [comment]
@Alys something that hasn't been suggested but is standard UX is collapsing notification messages. On any social media site you don't receive a line in your notifications per like but one for everyone who liked an item. We could place a threshold so 4 messages don't collapse but 5+ followed by a user message do. That way a party that hardly chats would still see their notifications.

**Comment by texastoland** [comment]
PS my party has _5 members_ and it interrupts conversation. I'm going to look into implementation next weekend.

**Comment by Alys** [comment]
@texastoland We do want to implement batch casting of skills (i.e., select a skill and select the number of times you want to cast it, with 1 being the default, and your mana limiting how large a number you can select of course). That would result in a single message in party chat for all casts, as well as being a significant improvement for usability in skill casting. I imagine the user interface could be similar to the bulk purchase interface (for eggs for example). I don't think we have an issue for that. My personal feeling is it would be better to work on that than to work on a way for collapsing individual messages, however I don't know how the staff feel about it (or you! :smile:).  You might like to comment here with a description of what you're thinking of implementing and see if there's any feedback.

**Comment by texastoland** [comment]
@Alys batching spells makes sense for usability but this would be a separate issue. I started writing up my proposal but it was rather long. Here's the outline:

1. Extract quest notifications from message boxes for differentiation and vertical flow.
2. Remove `code` tags and style them separately.
3. Replace text with visuals and numbers. Imagine an avatar icon, red and green numbers on either side for boss damage or quest items, followed by a series of spell icons, overlaid with number badges for the number of times cast. Wouldn't a screenshot be clearer? Same idea.
4. Collapse adjacent quest notifications. If text it could fade with an expand button like "see more" on Facebook. If icons they could be laid out horizontally.

Hopefully that's constructive.

**Comment by Alys** [comment]
@texastoland If the text were replaced, we'd need to ensure that the visuals were completely clear to all users. I'm not sure if that would be easy to achieve.

One thing I'm keen on is keeping it very obvious to anyone visiting the party page that there's been damage that affects them (we already get questions from inexperienced users about why they lost health because they don't realise the boss has damaged them and those questions would increase if boss damage wasn't as obvious). 

If you'd like to show us a mocked-up screenshot of your idea, that would be great!

**Comment by texastoland** [comment]
👍🏽 agreed all around @Alys. I'll send a grab before making changes.

**Comment by Tressley** [comment]
Hey @texastoland -- Just wanted to let you know that we've been doing some work around system messages and how they're displayed in chat. We're hoping these visual changes address a lot of the users' concerns around space and usability without needing to introduce new features or components just yet.

While these are upcoming changes, I'd encourage you to continue working on your proposal and sharing! We're always interested in seeing our contributor's ideas, as our features are always evolving and changing.

**Comment by texastoland** [comment]
Can't wait to see! I decided to tackle issues with a smaller scope.

**Comment by Alys** [comment]
Just to update this, there's [a staff comment above](https://github.com/HabitRPG/habitica/issues/3553#issuecomment-344718417) saying that batch / bulk skill casting is desired, but that situation has changed now as [noted here](https://github.com/HabitRPG/habitica/issues/10685#issuecomment-608541471) by @veeeeeee (redphoenix).

**Comment by Alys** [comment]
For anyone wanting this feature, there's a third-party style that hides party system messages: [Hide System Messages in Party on Habitica](https://habitica.fandom.com/wiki/Hide_System_Messages_in_Party_on_Habitica)

**Comment by Alys** [comment]
For anyone wanting this feature, there's a third-party style that hides party system messages: [Hide System Messages in Party on Habitica](https://habitica.fandom.com/wiki/Hide_System_Messages_in_Party_on_Habitica) created by @citrusella
